qiskit_experiments/library/tomography/fitters/cvxpy_utils.py

```diff
@@ -51,54 +64,168 @@ def decorated_func(*args, **kwargs):
+
+    Args:
+        dim: The dimension of the complex square matrix.
+        hermitian: If True add constraint that the matrix is Hermitian.
+        psd: If True add a constraint that the matrix is positive
+             semidefinite.
+        trace: If True add constraint that the trace of the matrix is
+               the specified value.
+
+    Returns:
+        A tuple ``(mat.real, mat.imag, constraints)`` of two real CVXPY
+        matrix variables, and constraints.
+    """
+    mat_r = cvxpy.Variable((dim, dim))
+    mat_i = cvxpy.Variable((dim, dim))
+    cons = []
+
+    if hermitian:
+        cons += hermitian_constaint(mat_r, mat_i)
+    if trace is not None:
+        cons += trace_constaint(mat_r, mat_i, trace)
+    if psd:
+        cons += psd_constaint(mat_r, mat_i)
+    return mat_r, mat_i, cons
+
+
+def hermitian_constaint(mat_r: Variable, mat_i: Variable) -> List[Constraint]:
+    """Return CVXPY constraint for a Hermitian matrix variable.
+
+    Args:
+        mat_r: The CVXPY variable for the real part of the matrix.
+        mat_i: The CVXPY variable for the complex part of the matrix.
+
+    Returns:
+        A list of constraints on the real and imaginary parts.
+    """
+    return [mat_r == mat_r.T, mat_i == -mat_i.T]
+
+
+def psd_constaint(mat_r: Variable, mat_i: Variable) -> List[Constraint]:
+    """Return CVXPY Hermitian constraints for a complex matrix.
+
+    Args:
+        mat_r: The CVXPY variable for the real part of the matrix.
+        mat_i: The CVXPY variable for the complex part of the matrix.
+
+    Returns:
+        A list of constraints on the real and imaginary parts.
+    """
+    bmat = cvxpy.bmat([[mat_r, -mat_i], [mat_i, mat_r]])
+    return [bmat >> 0]
+
+
+def trace_constaint(mat_r: Variable, mat_i: Variable, trace: complex) -> List[Constraint]:
+    """Return CVXPY trace constraints for a complex matrix.
+
+    Args:
+        mat_r: The CVXPY variable for the real part of the matrix.
+        mat_i: The CVXPY variable for the complex part of the matrix.
+        trace: The value for the trace constraint.
+
+    Returns:
+        A list of constraints on the real and imaginary parts.
+    """
+    return [cvxpy.trace(mat_r) == cvxpy.real(trace), cvxpy.trace(mat_i) == cvxpy.imag(trace)]
+
+
+def trace_preserving_constaint(mat_r: Variable, mat_i: Variable) -> List[Constraint]:
+    """Return CVXPY trace preserving constraints for a complex matrix.
+
+    Args:
+        mat_r: The CVXPY variable for the real part of the matrix.
+        mat_i: The CVXPY variable for the complex part of the matrix.
+
+    Returns:
+        A list of constraints on the real and imaginary parts.
+    """
+    sdim = int(np.sqrt(mat_r.shape[0]))
+    ptr = partial_trace_super(sdim, sdim)
+    return [
+        ptr @ cvxpy.vec(mat_r) == np.identity(sdim).ravel(),
+        ptr @ cvxpy.vec(mat_i) == np.zeros(sdim * sdim),
+    ]
+
+
+@functools.lru_cache(3)
+def partial_trace_super(dim1: int, dim2: int) -> np.array:
```

Comment:
What would be the situation we need to set different values for dim1 and dim2? Based on what is written here this is too much general? Would it be possible to simplify this when dim1==dim2? I got impression this is performance bottleneck since you set lru cache for this.

Reply:
Choi matrix / Channel from N qubits to M qubits (ie you do preparations on 1 qubits, but measurements on 2 qubits)